File: src/bin/2050.rs
struct Solution {}

impl Solution {
    pub fn minimum_time(n: i32, mut relations: Vec<Vec<i32>>, time: Vec<i32>) -> i32 {
        let n = n as usize;
        // 邻接表，但是前向
        let adj = {
            let mut adj: Vec<Vec<usize>> = vec![vec![]; n];
            for r in relations.iter_mut() {
                r[0] -= 1;
                r[1] -= 1;
                adj[r[1] as usize].push(r[0] as usize);
            }
            adj
        };
        // 由于 time[i] >= 0，所以 0 可以标记该值无效
        let mut mem = vec![0; n];

        fn eval(i: usize, mem: &mut Vec<i32>, adj: &Vec<Vec<usize>>, time: &Vec<i32>) -> i32 {
            if mem[i] != 0 {
                return mem[i];
            }
            let mut max_time = 0;
            for &course in adj[i].iter() {
                max_time = max_time.max(eval(course, mem, adj, time));
            }
            let ret = max_time + time[i];
            mem[i] = ret;
            ret
        }

        (0..n).map(|i| eval(i, &mut mem, &adj, &time)).max().unwrap()
    }
}

fn main() {}

File: src/cpp/2050.cpp
#include <cstdint>
#include <vector>

using namespace std;

class Solution {
    vector<int> course_min_time;
    vector<vector<int>> adj;

    int get_course_min_time(int idx, vector<int> &time) {
        if (course_min_time[idx] != 0) {
            return course_min_time[idx];
        }

        course_min_time[idx] = time[idx];
        int additional_time = 0;
        for (auto i : adj[idx]) {
            additional_time = max(additional_time, get_course_min_time(i, time));
        }
        course_min_time[idx] += additional_time;
        return course_min_time[idx];
    }

public:
    int minimumTime(int n, vector<vector<int>> &relations, vector<int> &time) {
        course_min_time.resize(n, 0);
        adj.resize(n);

        for (auto &v : relations) {
            adj[v[1] - 1].push_back(v[0] - 1);
        }

        int ret = 0;
        for (int i = 0; i < n; i++) {
            ret = max(ret, get_course_min_time(i, time));
        }

        return ret;
    }
};

int main() {
    return 0;
}